### PR TITLE
IPPM_plotter visual changes and refactoring

### LINKED ISSUE #224: Toolbox IPPM plots should look more like Kymata.org IPPM plots
E.g.
- Adds function names instead of legend (see #370)
- ~Adds option to print function equation instead~ → moved to #369 
- ~Adds better y coordinate~ → moved to #363 
- ~Adds special case for short edges~  → moved to #368 
- Adds arrowhead

### LINKED ISSUE #352: remove `ippm_parameter_evaluation` folder from demos
Hi @anirudh1666 and @caiw - I got Kymata to upload automatically to PyPI every time we tag a commit as a release (https://pypi.org/project/kymata/). However, I had to move Anirudh's 2 notebooks to a new folder - `demos/ippm_parameter_evaluation`, otherwise, they are added to the build. (Because they contain lots of PNG's, they would add considerably to the file size if I hadn't moved them.)

But them being in `demos` isn't quite right either as their is already a notebook in here that intends to demonstrate how to create an IPPM from the demo data.

Could you both discuss together how best to handle them? They should perhaps be combined and be relabelled something like `demo_for_automatic_IPPM_evaluation`.

### LINKED ISSUE #360: 70% of CI testing time is spent running `test_download_and_delete_fsaverage_surfaces`
This is possibly also violating some TOS of MNE/OSF, if we're downloading multiple MB of data on each run of the CI.

Consider deleting this test?

Or have a separate CI runner which runs once a day/week on a cron job rather than on every commit to a live PR?

### LINKED ISSUE #364: IPPM: Option for more than one input node
Currently only one is an option, but in future we will have (e.g.) left/right inputs

### LINKED ISSUE #366: Update version of matplotlib
The current version of matplotlib in the lockfile is 3.9.1, which has been [yanked](https://github.com/matplotlib/matplotlib/issues/28551). We should update.

### LINKED ISSUE #368: When IPPM edges are too short they cause zig-zags

### LINKED ISSUE #370: Display function names on IPPM plots
At the moment it just displays literally `function_name()`

### PR DESCRIPTION
**Pull request by neukym**
Aims to:
- [x] Fixes #224
  - [x] Adds function names to edges instead of legend
  - [x] Adds arrowhead
  - [x] Fixes #368 
  - [x] Make the nodes into little hexagons like on kymata.org
- [x] Fixes #352 
- [x] Fixes #360  
- [x] Can now merge hemispheres into one before clustering to double datapoints. If the hemispheres behave the same, then this will make results far more robust. One con is that the hyperparameter settings will have to be adjusted to reflect the greater density of data, e.g., base_bin_size for AMP. If hemispheres are merged, the final output is written to the hemisphere specified in the "hemi" hyperparameter.
- [x] Shuffling removes spurious correlations between rows. Consequently, the IPPM will be able to generalize easier. Although expression plots are time-series data, we don't actually analyze auto-correlation or any other temporal features. In one of my experiments, I will analyze the impact of incorporating a history of the last X magnitudes as an additional feature.
- [x] Fixes #366 
- [x] Fixes #370 

Stretch-goals:
- [x] Fixes #364 
- [x] #363 
- [x] #369 

Does not cover, but relates to:
- #362
- #367 

## More detailed changes, with motivation

- **Rename clusterers and strategies**: When I was reading through the code, I was confused where, for example, there was a `class CustomGMM(CustomClusterer)` and a `sklearn.cluster.GMM` in close proximity, or a `sklearn.cluster.DBSCAN` and a `class DBSCAN(DenoisingStrategy)` — it wasn't always clear just from reading the usages what roles were being fulfilled. So I renamed all clusterers to be `*Clusterer`, and renamed all strategies to be `*Strategy`. I also shortened the filenames from `custom_clusterers.py` to `cluster.py`.
- **Using constants to name hemispheres**: there are a lot of places where the strings `"leftHemisphere"` and `"rightHemisphere"` were being written as literals in the code, which always feels risky for a maintainance point of view. Since we also have other places where we name hemspheres using strings, I've set one place for the names in `constants.py`, and reused them everywhere.
- **Renamed `Node(NamedTuple)` to `IPPMNode(NamedTuple)`**: this was copied into multiple files, so I just replaced with a single instance and update usages.
- **Renamed `IPPMHexel` to `IPPMSpike`**: This was used both for actual hexels (i.e. pre-denoising) and the collapsed peaks (after denoising). Also in theory IPPMs could be generated from sensor data, which has sensors rather than hexels. So I changed the name to be more general.
- **Wraps `DBSCAN` and `MeanShift` into `CustomClusterers**: to reduce deep coupling with sklearn and keep the `CustomClusterer` abstract class as the true superclass.

### CODE REVIEW

**Comment by caiw** [comment]
A lot of the changes merged to `main` as part of #309 and #322 introduce conflicts with this branch, including moved files and things which are a pain to resolve. I think I'll just remake my changes on top and then force-push the new branch to this upstream to continue with this PR.

**Comment by caiw** [comment]
Rebased over `main`

**Comment by anirudh1666** [comment]
Updated GMM to not exclude anomalies since I did not like the strategy we used. Also added a check for singular covariance matrix, which indicates we should rerun the GMM algorithm.

Updated unit and integration tests to reflect new functionality.

**Comment by anirudh1666** [comment]
New CR adds functionality:
- Can now merge hemispheres into one before clustering to double datapoints. If the hemispheres behave the same, then this will make results far more robust. One con is that the hyperparameter settings will have to be adjusted to reflect the greater density of data, e.g., base_bin_size for AMP. If hemispheres are merged, the final output is written to the hemisphere specified in the "hemi" hyperparameter.
- Shuffling removes spurious correlations between rows. Consequently, the IPPM will be able to generalize easier. Although expression plots are time-series data, we don't actually analyze auto-correlation or any other temporal features. In one of my experiments, I will analyze the impact of incorporating a history of the last X magnitudes as an additional feature.

**Comment by caiw** [comment]
@neukym I've moved out some of the "nice-to-have" things into separate issues because this PR is too big already! I'll tackle those as separate smaller PRs when they become a priority.

**Comment by caiw** [comment]
@neukym I can't request you as a reviewer because you opened the PR, but please review this! :)

**Comment by neukym** [comment]
Looks good! Thanks both for some great work. Accepting and merging.
